gcc/rust/expand/rust-macro-builtins.cc

```diff
@@ -376,6 +390,29 @@ AST::Fragment
 MacroBuiltin::include_str_handler (Location invoc_locus,
 				   AST::MacroInvocData &invoc)
 {
+  auto invoc_token_tree = invoc.get_delim_tok_tree ();
+  MacroInvocLexer lex (invoc_token_tree.to_token_stream ());
+  Parser<MacroInvocLexer> parser (lex);
+
+  bool has_error = false;
+
+  auto last_token_id = macro_end_token (invoc_token_tree, parser);
+
+  auto expanded_expr = try_expand_many_expr (parser, last_token_id,
+					     invoc.get_expander (), has_error);
+
+  if (has_error)
+    {
+      AST::Fragment::create_error ();
+    }
```

Comment:
`include_str_handler` calls into `parse_single_string_literal`, which already takes care of lexing and parsing the tokens. So by adding another parser here you're effectively doing the same work twice. I think what we should do instead is look at how to modify `parse_single_string_literal` to allow parsing macro invocations. If you look at the function:

```c++
std::unique_ptr<AST::LiteralExpr>
parse_single_string_literal (AST::DelimTokenTree &invoc_token_tree,
                             Location invoc_locus, MacroExpander *expander)
{
  MacroInvocLexer lex (invoc_token_tree.to_token_stream ());
  Parser<MacroInvocLexer> parser (lex);

  auto last_token_id = macro_end_token (invoc_token_tree, parser);

  std::unique_ptr<AST::LiteralExpr> lit_expr = nullptr;

  if (parser.peek_current_token ()->get_id () == STRING_LITERAL)
    {
      lit_expr = parser.parse_literal_expr ();
      parser.maybe_skip_token (COMMA);
      if (parser.peek_current_token ()->get_id () != last_token_id)
        {
          lit_expr = nullptr;
          rust_error_at (invoc_locus, "macro takes 1 argument");
        }
    }
  else if (parser.peek_current_token ()->get_id () == last_token_id)
    rust_error_at (invoc_locus, "macro takes 1 argument");
  else
    rust_error_at (invoc_locus, "argument must be a string literal");

  parser.skip_token (last_token_id);

  return lit_expr;
}
```

you can see that it calls into `parse_literal_expr()`. We can discuss in more details on Zulip on how to go about making these modifications if you'd like, but I think that the first step would be to change its return type from `std::unique_ptr<AST::LiteralExpr>` to `std::unique_ptr<AST::Expr>` and trying to parse a literal expression or a macro invocation.

So I think the issue is a bit more complex than it looks but it's also a great way to learn about our parser and more of the codebase :)